File: 2016/src/day19.go
package src

import (
    "fmt"
    "strconv"
    "container/ring"
)

func Solve19(input string) string {
    var ans1 int
    var ans2 int

    ringSize, _ := strconv.Atoi(input)

    r := ring.New(ringSize)

    for i := 1; i <= ringSize; i++ {
        r.Value = i
        r = r.Next()
    }

    for i := 0; i < ringSize - 1; i++ {
        r.Unlink(1)
        r = r.Next()
    }

    ans1 = r.Value.(int)

    r = ring.New(ringSize)

    for i := 1; i <= ringSize; i++ {
        r.Value = i
        r = r.Next()
    }

    steps := (ringSize - 1) / 2 - (1 - ringSize % 2)

    for i := 0; i < steps; i++ {
        r = r.Next()
    }

    for ringSize > 0 {
        if ringSize % 2 == 0 {
            r = r.Next()
            r.Unlink(1)
        } else {
            r.Unlink(1)
        }
        ringSize--
    }

    ans2 = r.Value.(int)

    return fmt.Sprintf("%d, %d", ans1, ans2)
}

File: 2017/src/day19.js
export default function solve19(input) {
    let ans1 = "";
    let ans2 = 0;

    let grid = input.split("\n");

    let pos = [0, 0];
    grid[0].split("").forEach((c, i) => {
        if (c === "|") {
            pos[0] = i;
        }
    });

    let dir = [0, 1];

    while (dir !== null) {

        ans2++;

        let [dx, dy] = dir;
        pos[0] += dx;
        pos[1] += dy;

        let [x, y] = pos;
        let c = grid[y][x];

        if (c === "+") {
            dir = null;
            for (const [ox, oy] of [[-1, 0], [0, -1], [1, 0], [0, 1]]) {
                if (ox === -dx && oy === -dy) {
                    continue;
                }
                if (grid[y + oy][x + ox] !== " ") {
                    dir = [ox, oy];
                    break;
                }
            }
        } else if (c === " ") {
            break;
        } else if (c !== "-" && c !== "|") {
            ans1 += c;
        }
    }

    return [ans1, ans2];
}
